src/view-backend-private.cpp

```diff
@@ -103,16 +93,8 @@ void ViewBackend::exportEGLStreamProducer(struct wl_resource* bufferResource)
 
 void ViewBackend::dispatchFrameCallbacks()
 {
-    if (G_UNLIKELY(!m_client))
-        return;
-
-    FrameCallbackResource* resource;
-    wl_list_for_each(resource, &m_frameCallbacks, link) {
-        wl_callback_send_done(resource->resource, 0);
-    }
-    clearFrameCallbacks();
-
-    wl_client_flush(m_client);
+    if (G_LIKELY(m_surfaceId))
+        WS::Instance::singleton().dispatchFrameCallbacks(m_surfaceId);
```

Comment:
This is losing the `m_client` check. Please verify that the `m_surfaceId` check is a sufficient alternative.

Reply:
When a `wl_client` is destroyed all the resources associated with that client (including frame callbacks) are destroyed as well, which means the resource destruction callback introduced by this patch will run and remove all the callback resources from the `Surface::m_pendingFrameCallbacks` and `Surface::m_currentFrameCallbcks` lists:

```c
        wl_resource_set_implementation(callbackResource, nullptr, nullptr,
            [](struct wl_resource* resource) {
                wl_list_remove(wl_resource_get_link(resource));
        });
```

This means that if we reach `Surface::dispatchFrameCallbacks()` _after_ a certain `wl_client` has been destroyed, the callback resources would have already been removed from the lists.

Additionally, the `wl_callback_send_done()` calls from `Surface::dispatchFrameCallbacks()` end up in a call to `wl_resource_post_event()` which, looking in `wayland-server.c`, does a few sanity checks as well.

All in all, I think the approach suggested for handling the surface frame callback resources in this patch is as safe as it can be 🤔

Reply:
:+1: